tests: fix kernel panic on kvm switch tests

clemdiep opened the pull request:
The resource used for kvm switch tests as kernel arguments set to force the "root=/dev/sda2" ([x86-ubuntu-24.04-boot-with-systemd](https://resources.gem5.org/resources/x86-ubuntu-24.04-boot-with-systemd/raw?database=gem5-resources&version=5.0.0)). The kvm switch test adds some kernel arguments, but use the X86Board default kernel arguments. Which leads to "root=/dev/hda1" and a kernel panic occurs as it cannot mount root fs.

This commits fix that by getting the kernel arguments from the workload instead of the default ones.

clemdiep committed:
tests: fix kernel panic on kvm switch tests

The resource used for kvm switch tests as kernel arguments set to force
the "root=/dev/sda2". The kvm switch adds some kernel arguments, but
use the X86Board default kernel arguments. The argument is then set to
"root=/dev/hda1" and a kernel panic occurs as it cannot mount root fs.

This commits fix that by getting the kernel arguments from the workload
instead of the default ones.

mattsinc commented:
@abmerop , @v-ramadas : just checking if we need to integrate this with GPUFS?

powerjg reviewed (commented):
I feel like this should be fixed in `set_kernel_disk_workload` if the workload's kernel arguments aren't getting passed to the board.

Maybe we only need to remove line 166?

cc @Harshil2107

Harshil2107 commented on tests/gem5/kvm_switch_tests/configs/boot_kvm_switch_exit.py:
```diff
@@ -163,13 +163,20 @@
     resource_version="5.0.0",
 )
-workload.set_parameter("kernel_args", kernal_args)
+
+if "kernel_args" in workload.get_parameters():
+    kernel_args = workload.get_parameters()["kernel_args"]
+else:
+    kernel_args = motherboard.get_default_kernel_args()
+
+kernel_args += [args.kernel_args]
+
+workload.set_parameter("kernel_args", kernel_args)
 # Set the Full System workload.
 motherboard.set_workload(workload)
```
This change should fix the issue, but I believe we should be using the `append_kernel_arg` function instead. Boards that inherit from `KernelDiskWorkload` provide this function, which is intended to append kernel arguments to the existing ones. In our case, it should allow appending the arguments passed via argparse to the current `kernel_args` in the workload.

I recall there was a bug where `append_kernel_arg` wasn't actually appending the arguments. That issue was reported for RISC-V and appears to be fixed in #1936 . I'm currently double-checking whether it works as intended for x86 and Arm as well.

```suggestion
# Set the Full System workload.
motherboard.set_workload(workload)
motherboard.append_kernel_arg(args.kernel_args)
```

Harshil2107 replied:
The `append_kernel_arg` function seems to work as intended—it correctly adds the passed arguments to the `kernel_args` defined in the workload. I tested the change mentioned above by running the quick KVM switch tests, and they pass.

abmerop commented:
> @abmerop , @v-ramadas : just checking if we need to integrate this with GPUFS?

No, we don't use resources

erin-le commented:
@clemdiep: Hello, just letting you know that I'm planning to edit this PR to use Harshil's suggested changes.

erin-le committed:
tests: simplify adding kernel arguments in kvm switch tests

This commit modifies the kvm switch tests to use append_kernel_arg
to add the kernel arguments passed into the config via argparse.
This simplifies the logic in the config.

Harshil2107 reviewed (approved):
lgtm
